codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -315,12 +315,66 @@ void window(int initialWindowSize) {
 
         @Override
         int writeAllocatedBytes(int allocated) {
+            final int initialAllocated = allocated;
+            int writtenBytes;
+            // In case an exception is thrown we want to remember it and pass it to cancel(Throwable).
+            Throwable cause = null;
+            FlowControlled frame;
             try {
-                // Perform the write.
-                return writeBytes(allocated);
+                assert !writing;
+                writing = true;
+
+                // Write the remainder of frames that we are allowed to
+                boolean writeOccurred = false;
+                while (!cancelled && (frame = peek()) != null) {
+                    int maxBytes = min(allocated, writableWindow());
+                    if (maxBytes <= 0 && frame.size() > 0) {
+                        // The frame still has data, but the amount of allocated bytes has been exhausted.
+                        // Don't write needless empty frames.
+                        break;
+                    }
+                    writeOccurred = true;
```

Comment:
Just to see what it might look like I sketched up the following.  One difference is that if an exception is thrown the `allocated` may not be decremented the same way (about to get off train though :smile:)

```diff
diff --git a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java b/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java
index 674ae0e..7cad233 100644
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java
@@ -313,6 +313,20 @@ public class DefaultHttp2RemoteFlowController implements Http2RemoteFlowControll
             window = initialWindowSize;
         }
 
+        private int doWrite(FlowControlled frame, int maxBytes) {
+            int initialFrameSize = frame.size();
+            frame.write(ctx, max(0, maxBytes));
+            if (frame.size() == 0) {
+                // This frame has been fully written, remove this frame and notify it.
+                // Since we remove this frame first, we're guaranteed that its error
+                // method will not be called when we call cancel.
+                pendingWriteQueue.remove();
+                frame.writeComplete();
+            }
+            // Decrement allocated by how much was actually written.
+            return initialFrameSize - frame.size();
+        }
+
         @Override
         int writeAllocatedBytes(int allocated) {
             final int initialAllocated = allocated;
@@ -324,8 +338,15 @@ public class DefaultHttp2RemoteFlowController implements Http2RemoteFlowControll
                 assert !writing;
                 writing = true;
 
-                // Write the remainder of frames that we are allowed to
-                boolean writeOccurred = false;
+                if (!cancelled && (frame = peek()) != null) {
+                    int maxBytes = min(allocated, writableWindow());
+                    if (maxBytes <= 0 && frame.size() > 0) {
+                        // The frame still has data, but the amount of allocated bytes has been exhausted.
+                        // Don't write needless empty frames.
+                        return -1;
+                    }
+                    allocated -= doWrite(frame, maxBytes);
+                }
                 while (!cancelled && (frame = peek()) != null) {
                     int maxBytes = min(allocated, writableWindow());
                     if (maxBytes <= 0 && frame.size() > 0) {
@@ -333,28 +354,8 @@ public class DefaultHttp2RemoteFlowController implements Http2RemoteFlowControll
                         // Don't write needless empty frames.
                         break;
                     }
-                    writeOccurred = true;
-                    int initialFrameSize = frame.size();
-                    try {
-                        frame.write(ctx, max(0, maxBytes));
-                        if (frame.size() == 0) {
-                            // This frame has been fully written, remove this frame and notify it.
-                            // Since we remove this frame first, we're guaranteed that its error
-                            // method will not be called when we call cancel.
-                            pendingWriteQueue.remove();
-                            frame.writeComplete();
-                        }
-                    } finally {
-                        // Decrement allocated by how much was actually written.
-                        allocated -= initialFrameSize - frame.size();
-                    }
+                    allocated -= doWrite(frame, maxBytes);
                 }
-
-                if (!writeOccurred) {
-                    // Either there was no frame, or the amount of allocated bytes has been exhausted.
-                    return -1;
-                }
-
             } catch (Throwable t) {
                 // Mark the state as cancelled, we'll clear the pending queue via cancel() below.
                 cancelled = true;
```